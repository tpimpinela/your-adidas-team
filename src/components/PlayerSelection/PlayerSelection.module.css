.player-selection-container {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(180px, 1fr));
  gap: 24px;
}

.team-name {
  display: grid;
  place-items: center;
  color: white;
  margin-bottom: 2rem;
}
