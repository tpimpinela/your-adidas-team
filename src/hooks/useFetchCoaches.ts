import { useMemo } from "react";
import {
  CoachMapped,
  FootbalCoachesApiResponse,
} from "../models/coaches.models";
import { PlayerPosition } from "../models/squads.models";
import { API_ENDPOINTS } from "../utils/config";
import useFetch from "./useFetch";

const useFetchCoaches = (teamId: number) => {
  const { data, isLoading } = useFetch<FootbalCoachesApiResponse>(
    API_ENDPOINTS.coaches,
    {
      team: teamId,
    }
  );

  const mappedCoaches: CoachMapped[] = useMemo(() => {
    return (
      data?.response.map(({ id, name, firstname, lastname, photo, age }) => ({
        id,
        name,
        firstname,
        lastname,
        photo,
        age,
        teamId,
        position: PlayerPosition.Coach,
      })) ?? []
    );
  }, [data]);

  return {
    coaches: mappedCoaches,
    isLoading,
  };
};

export default useFetchCoaches;
