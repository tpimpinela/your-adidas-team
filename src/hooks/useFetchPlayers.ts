import { useMemo } from "react";
import {
  FootbalSquadsApiResponse,
  PlayerMapped,
} from "../models/squads.models";
import { API_ENDPOINTS } from "../utils/config";
import useFetch from "./useFetch";

const useFetchPlayers = (teamId: number) => {
  const { data, isLoading } = useFetch<FootbalSquadsApiResponse>(
    API_ENDPOINTS.squads,
    {
      team: teamId,
    }
  );

  const teamName = useMemo(() => data?.response[0].team.name, [data]);

  const mappedPlayers: PlayerMapped[] = useMemo(() => {
    return (
      data?.response[0].players.map(
        ({ id, name, number, position, photo }) => ({
          id,
          name,
          number,
          position,
          photo,
          teamId,
        })
      ) ?? []
    );
  }, [data]);

  return {
    players: mappedPlayers,
    teamName,
    isLoading,
  };
};

export default useFetchPlayers;
