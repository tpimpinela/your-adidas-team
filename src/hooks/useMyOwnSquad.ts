import { useCallback, useContext } from "react";
import { MyOwnSquadContext } from "../contexts/MyOwnSquad.context";

const useMyOwnSquad = () => {
  const myOwnSquad = useContext(MyOwnSquadContext);

  const isMemberOnMySquad = useCallback(
    (teamMemberId: number) => {
      return Boolean(myOwnSquad.find(({ id }) => id === teamMemberId));
    },
    [myOwnSquad]
  );

  return {
    squad: myOwnSquad,
    isMemberOnMySquad,
  };
};

export default useMyOwnSquad;
