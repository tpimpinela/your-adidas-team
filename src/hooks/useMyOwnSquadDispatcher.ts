import { useCallback, useContext } from "react";
import { MyOwnSquadDispatchContext } from "../contexts/MyOwnSquad.context";
import { TeamMember } from "../models/teamMember.model";
import { MyOwnSquadActionType } from "../reducers/myOwnSquad.reducer";

const useMyOwnSquadDispatcher = () => {
  const dispatch = useContext(MyOwnSquadDispatchContext);

  const addTeamMember = useCallback(
    (teamMember: TeamMember) => {
      dispatch({
        type: MyOwnSquadActionType.Add,
        payload: teamMember,
      });
    },
    [dispatch]
  );

  const deleteTeamMember = useCallback(
    (id: number) => {
      dispatch({
        type: MyOwnSquadActionType.Delete,
        payload: id,
      });
    },
    [dispatch]
  );

  return {
    addTeamMember,
    deleteTeamMember,
  };
};

export default useMyOwnSquadDispatcher;
