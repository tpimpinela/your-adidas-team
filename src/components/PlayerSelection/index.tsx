import { useCallback, useMemo } from "react";
import { useNavigate, useParams } from "react-router-dom";
import useFetchCoaches from "../../hooks/useFetchCoaches";
import useFetchPlayers from "../../hooks/useFetchPlayers";
import useMyOwnSquad from "../../hooks/useMyOwnSquad";
import useMyOwnSquadDispatcher from "../../hooks/useMyOwnSquadDispatcher";
import { TeamMember as TeamMemberModel } from "../../models/teamMember.model";
import { MAX_PLAYERS_PER_TEAM } from "../../utils/config";
import Button from "../Button";
import TeamMember from "../TeamMember";
import styles from "./PlayerSelection.module.css";

const PlayerSelection = () => {
  const { teamId } = useParams() as { teamId: string };
  const { players, teamName } = useFetchPlayers(parseInt(teamId));
  const { coaches } = useFetchCoaches(parseInt(teamId));
  const { addTeamMember, deleteTeamMember } = useMyOwnSquadDispatcher();
  const {
    validations: { disabledTeamIds, maxPlayers },
  } = useMyOwnSquad();
  const navigate = useNavigate();

  const handleTeamMemberClick = useCallback(
    (teamMember: TeamMemberModel, isOnTheSquad: boolean) => {
      if (isOnTheSquad) {
        deleteTeamMember(teamMember.id);
      } else if (!isTeamIdDisabled) {
        addTeamMember(teamMember);
      }
    },
    [addTeamMember]
  );

  const handleSeeYourTeam = useCallback(() => {
    navigate("/");
  }, [navigate]);

  const isTeamIdDisabled = useMemo(
    () => disabledTeamIds?.includes(parseInt(teamId)),
    [disabledTeamIds, teamId]
  );

  return (
    <>
      <div className={styles.actions}>
        <Button onClick={handleSeeYourTeam}>See your team</Button>
      </div>
      <div className={styles.headings}>
        <h2 className={styles["headings__name"]}>{teamName}</h2>
        {(isTeamIdDisabled || maxPlayers) && (
          <h3
            data-testid="warning-header"
            className={styles["headings__warning"]}
          >
            {maxPlayers
              ? `You have reached the maximum number of players on your team.`
              : `You can't add more players from this team. Please, deselect one
                or check another team.`}
          </h3>
        )}
      </div>
      <div className={styles["player-selection-container"]}>
        {[...(players ?? []), ...(coaches ?? [])]?.map((teamMember) => (
          <TeamMember
            onTeamMemberClicked={handleTeamMemberClick}
            teamMember={teamMember}
            disabled={isTeamIdDisabled || Boolean(maxPlayers)}
            key={teamMember.id}
          />
        ))}
      </div>
    </>
  );
};

export default PlayerSelection;
