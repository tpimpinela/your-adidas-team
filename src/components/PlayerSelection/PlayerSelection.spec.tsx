import { render, RenderResult } from "@testing-library/react";
import { useNavigate, useParams } from "react-router-dom";
import { beforeEach, describe, expect, it, Mock, vi } from "vitest";
import PlayerSelection from ".";
import useFetchCoaches from "../../hooks/useFetchCoaches";
import useFetchPlayers from "../../hooks/useFetchPlayers";
import coachMock from "../../mocks/coach.mock";
import squadMock from "../../mocks/squads.mock";
import { TeamMember } from "../../models/teamMember.model";
import userEvent from "@testing-library/user-event";
import useMyOwnSquadDispatcher from "../../hooks/useMyOwnSquadDispatcher";
import useMyOwnSquad from "../../hooks/useMyOwnSquad";

vi.mock("react-router-dom", () => ({
  useParams: vi.fn(),
  useNavigate: vi.fn(),
}));
vi.mock("../../hooks/useFetchCoaches", () => ({
  default: vi.fn(),
}));
vi.mock("../../hooks/useFetchPlayers", () => ({
  default: vi.fn(),
}));
vi.mock("../TeamMember", () => ({
  default: ({
    onTeamMemberClicked,
  }: {
    onTeamMemberClicked: (
      teamMember: TeamMember,
      isOnTheSquad: boolean
    ) => void;
  }) => (
    <>
      <h1 data-testid="team-member">TeamMember</h1>
      <button
        data-testid="click-true"
        onClick={() => {
          onTeamMemberClicked({} as any, true);
        }}
      >
        Click me
      </button>
      <button
        data-testid="click-false"
        onClick={() => {
          onTeamMemberClicked({} as any, false);
        }}
      >
        Click me
      </button>
    </>
  ),
}));
vi.mock("../../hooks/useMyOwnSquadDispatcher", () => ({
  mock1: vi.fn(),
  mock2: vi.fn(),
  default() {
    return {
      addTeamMember: this.mock1,
      deleteTeamMember: this.mock2,
    };
  },
}));
vi.mock("../../hooks/useMyOwnSquad", () => ({
  default: vi.fn(),
}));

describe("<PlayerSelection />", () => {
  let renderResult: RenderResult;
  const teamIdMock = 9;
  const useParamsMock = useParams as Mock<any, any>;
  const useFetchCoachesMock = useFetchCoaches as Mock<any, any>;
  const useFetchPlayersMock = useFetchPlayers as Mock<any, any>;
  const addTeamMemberMock = useMyOwnSquadDispatcher().addTeamMember as Mock<
    any,
    any
  >;
  const deleteTeamMemberMock = useMyOwnSquadDispatcher()
    .deleteTeamMember as Mock<any, any>;
  const useMyOwnSquadMock = useMyOwnSquad as Mock<any, any>;
  let useNavigateMock: Mock<any, any>;

  beforeEach(() => {
    useNavigateMock = vi.fn();
    (useNavigate as Mock<any, any>).mockReturnValue(useNavigateMock);
    useMyOwnSquadMock.mockReturnValue({
      validations: {},
    });
    useParamsMock.mockReturnValue(teamIdMock);
    useFetchCoachesMock.mockReturnValue({
      coaches: [coachMock],
      isLoading: false,
    });
    useFetchPlayersMock.mockReturnValue({
      players: squadMock.players,
      isLoading: false,
      teamName: squadMock.team.name,
    });
    renderResult = render(<PlayerSelection />);
  });

  it("should match the snapshot", () => {
    expect(renderResult.container).toMatchSnapshot();
  });

  it("should navigate to root page if button is clicked", async () => {
    await userEvent.click(renderResult.getByTestId("button-component"));
    expect(useNavigateMock).toHaveBeenCalledWith("/");
  });

  it("should show team name", () => {
    expect(renderResult.getByRole("heading", { level: 2 }).textContent).toBe(
      squadMock.team.name
    );
  });

  it("should show one TeamMember component per player and coach", () => {
    expect(renderResult.getAllByTestId("team-member").length).toBe(2);
  });

  describe("on team member clicked", () => {
    it("should add team member if is not already on the squad", async () => {
      await userEvent.click(renderResult.getAllByTestId("click-false")[0]);
      expect(addTeamMemberMock).toHaveBeenCalled();
    });

    it("should delete team member if is already on the squad", async () => {
      await userEvent.click(renderResult.getAllByTestId("click-true")[0]);
      expect(deleteTeamMemberMock).toHaveBeenCalled();
    });
  });

  describe("if team is disabled or the squad has reached the maximum number of players", () => {
    it("should show a header with a warning", () => {
      useMyOwnSquadMock.mockReturnValue({
        validations: {
          maxPlayers: "test",
        },
      });
      renderResult.rerender(<PlayerSelection />);
      expect(renderResult.getByTestId("warning-header")).toBeTruthy();
    });

    it("should render correct text if the squad has reached the maximum number of players", () => {
      useMyOwnSquadMock.mockReturnValue({
        validations: {
          maxPlayers: "test",
        },
      });
      renderResult.rerender(<PlayerSelection />);
      expect(renderResult.getByTestId("warning-header").textContent).toBe(
        "You have reached the maximum number of players on your team."
      );
    });

    it("should render correct text if the squad has reached the maximum number of players of this team", () => {
      it("should render correct text if the squad has reached the maximum number of players", () => {
        useMyOwnSquadMock.mockReturnValue({
          validations: {
            disabledTeamIds: [teamIdMock],
          },
        });
        renderResult.rerender(<PlayerSelection />);
        expect(renderResult.getByTestId("warning-header").textContent).toBe(
          "You can't add more players from this team. Please, deselect one or check another team."
        );
      });
    });
  });
});
