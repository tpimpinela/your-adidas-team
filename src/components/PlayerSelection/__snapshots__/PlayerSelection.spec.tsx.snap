// Vitest Snapshot v1

exports[`<PlayerSelection /> > should match the snapshot 1`] = `
<div>
  <div
    class="_actions_e89e23"
  >
    <button
      class="_button_2a1efe"
      data-testid="button-component"
    >
      See your team
    </button>
  </div>
  <div
    class="_headings_e89e23"
  >
    <h2
      class="_headings__name_e89e23"
    >
      Spain
    </h2>
  </div>
  <div
    class="_player-selection-container_e89e23"
  >
    <h1
      data-testid="team-member"
    >
      TeamMember
    </h1>
    <button
      data-testid="click-true"
    >
      Click me
    </button>
    <button
      data-testid="click-false"
    >
      Click me
    </button>
    <h1
      data-testid="team-member"
    >
      TeamMember
    </h1>
    <button
      data-testid="click-true"
    >
      Click me
    </button>
    <button
      data-testid="click-false"
    >
      Click me
    </button>
  </div>
</div>
`;
